# Capstone Project

### Problem Description

#### The client wants to start the restaurant in Chennai in Tamil Nadu, India. He is looking to start a big Chinese restaurant, so he wants to build it in a main area in the city where there is more footfall. He also doesn’t want a place where there are many already an established Chinese restaurant. The goal of the project is to find is to find an area in Chennai which has most of the crowded places and also the least number of Chinese type restaurants. 

In [1]:
#to get location data using pincodes
pip install pgeocode

Note: you may need to restart the kernel to use updated packages.


In [2]:
#for maps
pip install folium

     |████████████████████████████████| 94 kB 6.8 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
import pgeocode
import folium
import pandas as pd
import numpy as np
import requests #request for Foursquare API
import json
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans #clustering algorithm
import matplotlib.cm as cm #for maps
import matplotlib.colors as colors #for maps

### Creating a dataframe which has the pincodes for all of the areas in Chennai.

In [4]:
read_page = pd.read_html('https://www.mapsofindia.com/pincode/india/tamil-nadu/chennai/')

In [5]:
pcodes_df = pd.DataFrame(data = read_page[0])
pcodes_df.columns = pcodes_df.iloc[0]
pcodes_df.head()

In [6]:
only_pincodes = pcodes_df[['Pincode']]
pincodes = only_pincodes.drop_na() #dropping rows because some of the pincodes didn't have values
pincodes.sort_values(by = 'Pincode')
pincodes.reset_index(inplace = True)

In [7]:
location_list = []

### Using the pincodes from above to get the areas and their associated latitude and longitudes using the pgeocode library

In [8]:
# The areas in Chennai are given a pincode. Some of the areas may have the same pincode, so they are combined as one area. 
# (like pincode 600018 for Pr. Accountant General and Teynampet)
for code in range(len(pincodes)):
    pin = pincodes.iloc[code][1]
    nomi = pgeocode.Nominatim('in')
    result = nomi.query_postal_code(pin)
    location_list.append([result.postal_code, result.place_name, result.latitude, result.longitude])

In [9]:
location_list[0:5] #list which contains all the details as a list

[['600018', 'Pr. Accountant General, Teynampet', 13.0433, 80.2528],
 ['600020',
  'Shastri Nagar (Chennai), Kasturibai Nagar, Adyar (Chennai)',
  12.9967,
  80.2603],
 ['600082', 'Periyar Nagar, Kumaran Nagar, G K M Colony', 13.0572, 80.2554],
 ['600029', nan, nan, nan],
 ['600040', 'Anna Nagar (Chennai)', 12.8819, 80.0885]]

In [10]:
places_with_location_df = pd.DataFrame(location_list, columns = ['Pincode','Places','Latitude','Longitude']) #converting the list into dataframe

In [11]:
places_with_location_df.head()

Pincode                                             Places  Latitude  \
0  600018                  Pr. Accountant General, Teynampet   13.0433   
1  600020  Shastri Nagar (Chennai), Kasturibai Nagar, Ady...   12.9967   
2  600082         Periyar Nagar, Kumaran Nagar, G K M Colony   13.0572   
3  600029                                                NaN       NaN   
4  600040                               Anna Nagar (Chennai)   12.8819   

   Longitude  
0    80.2528  
1    80.2603  
2    80.2554  
3        NaN  
4    80.0885

### Checking the data for empty values

In [12]:
chennai_data = places_with_location_df.dropna()

In [13]:
chennai_data.shape

(55, 4)

In [14]:
#Chennai city latitude and longitude to view in the map
chennai_lat = 13.067439
chennai_long = 80.237617

### Creating a map to view the areas within Chennai

In [15]:
chennai_map = folium.Map(location=[chennai_lat, chennai_long], zoom_start=10)
for lat, lng, places in zip(chennai_data['Latitude'], chennai_data['Longitude'], chennai_data['Places']):
    label = places
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(chennai_map)  
    
chennai_map

### Creating a Foursquare connection to get the venues in each area

In [16]:
CLIENT_ID = 'VS30EBSEKGFBYJP1OSNATOVVP5MUSKURQULLVXXGAPXPYJSM' # Foursquare ID 
CLIENT_SECRET = 'XGZTKYOOMT2QCNHBBBQBPAMBKBIHTHVECF11GGDIYRF35AQI' # Foursquare Secret 
ACCESS_TOKEN = 'FKUWDU1HQ4GENCZLJZ3BBLV1NNUOO24BMASLHM5RT5KIOXAH' # FourSquare Access Token
VERSION = '20180605' # API version

In [ ]:
### Exploring one area (Anna nagar) using the Foursquare API

In [17]:
# Creating url make request to the Foursquare API
# A limit of 100 is maitained for all the areas
LIMIT = 100
# lat and long of Anna nagar
latitude = 12.8819
longitude = 80.0885
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude,  
    LIMIT)

results = requests.get(url).json()

In [18]:
# function to retrieve the category of the venue from the json result
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### Converting the resulting venues of Anna nagar to a dataframe

In [19]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name                      categories        lat  \
0              Vandalur zoo                     Zoo Exhibit  12.879441   
1  Kumbakonam Degree Coffee                     Coffee Shop  12.833756   
2            ibaco selaiyur                  Ice Cream Shop  12.922280   
3            Domino's Pizza                     Pizza Place  12.926000   
4               Honey Spice  Multicuisine Indian Restaurant  12.870879   

         lng  
0  80.081328  
1  80.048721  
2  80.138609  
3  80.108000  
4  80.076680

### Creating the function to explore all the areas in Chennai

In [20]:
def getNearbyVenues(names, latitude, longitude):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitude, longitude):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng,
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Places', 
                  'Latitude of the place', 
                  'Longitude of the place', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
#running the above function
chennai_venues = getNearbyVenues(names=chennai_data['Places'],
                                   latitude=chennai_data['Latitude'],
                                   longitude=chennai_data['Longitude']
                                  )

Pr. Accountant General, Teynampet
Shastri Nagar (Chennai), Kasturibai Nagar, Adyar (Chennai)
Periyar Nagar, Kumaran Nagar, G K M Colony
Anna Nagar (Chennai)
Anna Nagar East
Anna Nagar Western Extn
Chintadripet, Anna Road H.O, Madras Electricity System
Arumbakkam, D G Vaishnav College
Jafferkhanpet, Ashoknagar (Chennai)
Aynavaram
Besantnagar, Rajaji Bhavan
Washermanpet, Washermanpet East
Tidel Park, TTTI Taramani
Ekkaduthangal, Guindy Industrial Estate
Sowcarpet, Mannady (Chennai), Chennai G.P.O., Flower Bazaar, Mint Building, Govt Stanley Hospital
Chepauk, Tiruvallikkeni, Parthasarathy Koil, Madras University
Chetput, World University Centre
Choolaimedu
Venkatesapuram, Puliyanthope, Perambur Barracks
Engineering College (Chennai)
Madras Medical College, Park Town H.O, Edapalayam, Ripon Buildings
Egmore, Ethiraj Salai, Egmore ND
Erukkancheri, Kodungaiyur, Rv Nagar
Flowers Road
Raja Annamalaipuram, Ramakrishna Nagar (Chennai)
Fort St George
Royapettah, Lloyds Estate
Gopalapuram (Chennai)

In [23]:
chennai_venues.shape #all the venues from all the areas in the dataset

(4570, 7)

In [24]:
chennai_venues.head()

Places  Latitude of the place  \
0  Pr. Accountant General, Teynampet                13.0433   
1  Pr. Accountant General, Teynampet                13.0433   
2  Pr. Accountant General, Teynampet                13.0433   
3  Pr. Accountant General, Teynampet                13.0433   
4  Pr. Accountant General, Teynampet                13.0433   

   Longitude of the place                         Venue  Venue Latitude  \
0                 80.2528                          Stix       13.042898   
1                 80.2528         Hyatt Regency Chennai       13.042874   
2                 80.2528                  Ente Keralam       13.042188   
3                 80.2528  Z The Tapas Bar & Restaurant       13.045808   
4                 80.2528                       Nando's       13.046237   

   Venue Longitude      Venue Category  
0        80.248640  Chinese Restaurant  
1        80.248593               Hotel  
2        80.255880   Kerala Restaurant  
3        80.258013                 Bar  
4        80.256267  African Restaurant

In [25]:
chennai_venues.groupby('Places').count() #to view how much of venues Foursquare has returned for each area

Latitude of the place  \
Places                                                                      
Anna Nagar (Chennai)                                                   39   
Anna Nagar East                                                       100   
Anna Nagar Western Extn                                               100   
Arumbakkam, D G Vaishnav College                                      100   
Aynavaram                                                             100   
Besantnagar, Rajaji Bhavan                                             30   
Chepauk, Tiruvallikkeni, Parthasarathy Koil, Ma...                    100   
Chetput, World University Centre                                      100   
Chintadripet, Anna Road H.O, Madras Electricity...                    100   
Choolaimedu                                                            30   
Egmore, Ethiraj Salai, Egmore ND                                      100   
Ekkaduthangal, Guindy Industrial Estate                               100   
Engineering College (Chennai)                                         100   
Erukkancheri, Kodungaiyur, Rv Nagar                                   100   
Flowers Road                                                          100   
Fort St George                                                        100   
Gopalapuram (Chennai)                                                 100   
High Court Building (Chennai)                                         100   
Icf Colony                                                            100   
Indian Institute Of Technology                                        100   
Jafferkhanpet, Ashoknagar (Chennai)                                   100   
Kalaignar Karunanidhi Nagar                                            40   
Kilpauk, Kilpauk Medical College                                       71   
Kodambakkam                                                            96   
Kotturpuram                                                           100   
Madras Medical College, Park Town H.O, Edapalay...                    100   
Mylapore H.O, Mandaveli, Vivekananda College Ma...                    100   
Nandanam                                                              100   
Nerkundram, Koyambedu Wholesale Market Com, Koy...                    100   
Nungambakkam, Nungambakkam High Road, Loyola Co...                    100   
Perambur, Sembiam, Perambur North                                      61   
Periyar Nagar, Kumaran Nagar, G K M Colony                            100   
Pr. Accountant General, Teynampet                                     100   
Raja Annamalaipuram, Ramakrishna Nagar (Chennai)                      100   
Rajbhavan (Chennai)                                                   100   
Rayapuram                                                              50   
Royapettah, Lloyds Estate                                              82   
Saidapet (Chennai), Guindy North                                       48   
Saligramam                                                            100   
Shastri Nagar (Chennai), Kasturibai Nagar, Adya...                    100   
Shenoy Nagar, Aminjikarai                                              85   
Sowcarpet, Mannady (Chennai), Chennai G.P.O., F...                     89   
Teynampet West, DPI, Shastri Bhavan, Greams Road                      100   
Thygarayanagar North ND, Thygarayanagar South N...                     48   
Tidel Park, TTTI Taramani                                              20   
Tondiarpet, Tondiarpet West, Tondiarpet Bazaar                         38   
Vadapalani                                                             82   
Valmiki Nagar, Tiruvanmiyur, Palavakkam (Kanchi...                     73   
Velacheri                                                             100   
Venkatesapuram, Puliyanthope, Perambur Barracks                        18   
Vepery                                          

### As the radius was not given in the url request section, there may be duplicate values in the dataset. So it must be cleaned.

In [27]:
chennai_venues.loc[chennai_venues['Venue'] == 'Subway'] #checking for duplicate values for a the venue Subway

Places  \
38                    Pr. Accountant General, Teynampet   
96                    Pr. Accountant General, Teynampet   
162   Shastri Nagar (Chennai), Kasturibai Nagar, Ady...   
175   Shastri Nagar (Chennai), Kasturibai Nagar, Ady...   
193   Shastri Nagar (Chennai), Kasturibai Nagar, Ady...   
...                                                 ...   
4200  Valmiki Nagar, Tiruvanmiyur, Palavakkam (Kanch...   
4275                                         Vadapalani   
4279                                         Vadapalani   
4348                                          Velacheri   
4372                                          Velacheri   

      Latitude of the place  Longitude of the place   Venue  Venue Latitude  \
38                  13.0433                 80.2528  Subway       13.054787   
96                  13.0433                 80.2528  Subway       13.049846   
162                 12.9967                 80.2603  Subway       12.989579   
175                 12.9967                 80.2603  Subway       12.999040   
193                 12.9967                 80.2603  Subway       13.017396   
...                     ...                     ...     ...             ...   
4200                12.9695                 80.2561  Subway       12.999040   
4275                13.0511                 80.2125  Subway       13.038853   
4279                13.0511                 80.2125  Subway       13.047275   
4348                13.0647                 80.2523  Subway       13.061403   
4372                13.0647                 80.2523  Subway       13.054787   

      Venue Longitude  Venue Category  
38          80.249618  Sandwich Place  
96          80.256958  Sandwich Place  
162         80.248830  Sandwich Place  
175         80.254743  Sandwich Place  
193         80.271303  Sandwich Place  
...               ...             ...  
4200        80.254743  Sandwich Place  
4275        80.212274  Sandwich Place  
4279        80.194957  Sandwich Place  
4348        80.248521  Sandwich Place  
4372        80.249618  Sandwich Place  

[85 rows x 7 columns]

### It's been confirmed that there are duplicate values for Subway. There must be many like this. The dataframe currently contains 4570 samples.

In [28]:
chennai_venues.drop_duplicates(subset = ['Places', 'Venue'], keep = 'last', inplace = True) #dropping duplicate values

In [29]:
chennai_venues.shape

(4252, 7)

In [30]:
chennai_venues.loc[chennai_venues['Venue'] == 'Subway']

Places  \
96                    Pr. Accountant General, Teynampet   
196   Shastri Nagar (Chennai), Kasturibai Nagar, Ady...   
288          Periyar Nagar, Kumaran Nagar, G K M Colony   
323                                Anna Nagar (Chennai)   
405                                     Anna Nagar East   
505                             Anna Nagar Western Extn   
603   Chintadripet, Anna Road H.O, Madras Electricit...   
639                    Arumbakkam, D G Vaishnav College   
822                 Jafferkhanpet, Ashoknagar (Chennai)   
915                                           Aynavaram   
956                          Besantnagar, Rajaji Bhavan   
1072            Ekkaduthangal, Guindy Industrial Estate   
1349                   Chetput, World University Centre   
1432                                        Choolaimedu   
1453    Venkatesapuram, Puliyanthope, Perambur Barracks   
1539                      Engineering College (Chennai)   
1624  Madras Medical College, Park Town H.O, Edapala...   
1755                   Egmore, Ethiraj Salai, Egmore ND   
1798                Erukkancheri, Kodungaiyur, Rv Nagar   
1946                                       Flowers Road   
2051   Raja Annamalaipuram, Ramakrishna Nagar (Chennai)   
2124                                     Fort St George   
2310                              Gopalapuram (Chennai)   
2406   Teynampet West, DPI, Shastri Bhavan, Greams Road   
2451                   Saidapet (Chennai), Guindy North   
2559                      High Court Building (Chennai)   
2599  Thygarayanagar North ND, Thygarayanagar South ...   
2687                                         Icf Colony   
2787                     Indian Institute Of Technology   
2862                        Kalaignar Karunanidhi Nagar   
2969                                        Kodambakkam   
3032                   Kilpauk, Kilpauk Medical College   
3199  Nerkundram, Koyambedu Wholesale Market Com, Ko...   
3333                                      Virugambakkam   
3437                       West Mambalam, Mambalam R.S.   
3556  Nungambakkam, Nungambakkam High Road, Loyola C...   
3659  Mylapore H.O, Mandaveli, Vivekananda College M...   
3856                          Shenoy Nagar, Aminjikarai   
3907                  Perambur, Sembiam, Perambur North   
4020                                Rajbhavan (Chennai)   
4079                                         Saligramam   
4200  Valmiki Nagar, Tiruvanmiyur, Palavakkam (Kanch...   
4279                                         Vadapalani   
4372                                          Velacheri   

      Latitude of the place  Longitude of the place   Venue  Venue Latitude  \
96                  13.0433               80.252800  Subway       13.049846   
196                 12.9967               80.260300  Subway       12.986004   
288                 13.0572               80.255400  Subway       13.058682   
323                 12.8819               80.088500  Subway       12.823322   
405                 13.0974               80.195000  Subway       13.094722   
505                 13.0974               80.195000  Subway       13.094722   
603                 13.0744               80.271400  Subway       13.058682   
639                 13.0734               80.206900  Subway       13.082455   
822                 13.0582               80.240000  Subway       13.069381   
915                 13.0484               80.247300  Subway       13.049846   
956                 12.8503               80.070700  Subway       12.823322   
1072                13.0659               80.236500  Subway       13.077176   
1349                13.0700               80.240800  Subway       13.077176   
1432                12.8503               80.070700  Subway       12.823322   
1453                13.0647               80.047500  Subway       13.038856   
1539                13.0484               80.247300  Subway       13.049846   
1624                13.0754               80.27370

In [31]:
chennai_venues.loc[chennai_venues['Venue'] == 'Subway'].count()

Places                    44
Latitude of the place     44
Longitude of the place    44
Venue                     44
Venue Latitude            44
Venue Longitude           44
Venue Category            44
dtype: int64

### As we can see the subway duplicates are removed and the number of total samples are much reduced. It confirms that all the duplicates are removed.

### The dataset will be encoded inorder to view the most common places in all areas of Chennai. 
### This is done because we need to find the amount of restaurants and other common places in each area, so that we can decide on which area to choose.

In [32]:
# one hot encoding
chennai_onehot = pd.get_dummies(chennai_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
chennai_onehot['Places'] = chennai_venues['Places'] 

# move neighborhood column to the first column
fixed_columns = [chennai_onehot.columns[-1]] + list(chennai_onehot.columns[:-1])
chennai_onehot = chennai_onehot[fixed_columns]

chennai_onehot.head()

Places  African Restaurant  Airport Lounge  \
0  Pr. Accountant General, Teynampet                   0               0   
1  Pr. Accountant General, Teynampet                   0               0   
2  Pr. Accountant General, Teynampet                   0               0   
3  Pr. Accountant General, Teynampet                   0               0   
4  Pr. Accountant General, Teynampet                   1               0   

   American Restaurant  Arcade  Arts & Crafts Store  Asian Restaurant  \
0                    0       0                    0                 0   
1                    0       0                    0                 0   
2                    0       0                    0                 0   
3                    0       0                    0                 0   
4                    0       0                    0                 0   

   Athletics & Sports  BBQ Joint  Bakery  ...  Thai Restaurant  Theater  \
0                   0          0       0  ...                0        0   
1                   0          0       0  ...                0        0   
2                   0          0       0  ...                0        0   
3                   0          0       0  ...                0        0   
4                   0          0       0  ...                0        0   

   Theme Park  Toy / Game Store  Train Station  Vegetarian / Vegan Restaurant  \
0           0                 0              0                              0   
1           0                 0              0                              0   
2           0                 0              0                              0   
3           0                 0              0                              0   
4           0                 0              0                              0   

   Video Store  Women's Store  Zoo  Zoo Exhibit  
0            0              0    0            0  
1            0              0    0            0  
2            0              0    0            0  
3            0              0    0            0  
4            0              0    0            0  

[5 rows x 149 columns]

In [33]:
chennai_onehot.shape

(4252, 149)

### The dataframe is grouped so that the most common place in that particular area has the highest number.

In [34]:
chennai_onehot_grouped = chennai_onehot.groupby('Places').mean().reset_index()
chennai_onehot_grouped.head()

Places  African Restaurant  \
0                                Anna Nagar (Chennai)            0.000000   
1                                     Anna Nagar East            0.000000   
2                             Anna Nagar Western Extn            0.000000   
3                    Arumbakkam, D G Vaishnav College            0.000000   
4                                           Aynavaram            0.010309   
5                          Besantnagar, Rajaji Bhavan            0.000000   
6   Chepauk, Tiruvallikkeni, Parthasarathy Koil, M...            0.000000   
7                    Chetput, World University Centre            0.000000   
8   Chintadripet, Anna Road H.O, Madras Electricit...            0.000000   
9                                         Choolaimedu            0.000000   
10                   Egmore, Ethiraj Salai, Egmore ND            0.000000   
11            Ekkaduthangal, Guindy Industrial Estate            0.000000   
12                      Engineering College (Chennai)            0.010309   
13                Erukkancheri, Kodungaiyur, Rv Nagar            0.000000   
14                                       Flowers Road            0.000000   
15                                     Fort St George            0.000000   
16                              Gopalapuram (Chennai)            0.010309   
17                      High Court Building (Chennai)            0.000000   
18                                         Icf Colony            0.000000   
19                     Indian Institute Of Technology            0.000000   
20                Jafferkhanpet, Ashoknagar (Chennai)            0.000000   
21                        Kalaignar Karunanidhi Nagar            0.000000   
22                   Kilpauk, Kilpauk Medical College            0.000000   
23                                        Kodambakkam            0.000000   
24                                        Kotturpuram            0.000000   
25  Madras Medical College, Park Town H.O, Edapala...            0.000000   
26  Mylapore H.O, Mandaveli, Vivekananda College M...            0.000000   
27                                           Nandanam            0.000000   
28  Nerkundram, Koyambedu Wholesale Market Com, Ko...            0.000000   
29  Nungambakkam, Nungambakkam High Road, Loyola C...            0.000000   
30                  Perambur, Sembiam, Perambur North            0.000000   
31         Periyar Nagar, Kumaran Nagar, G K M Colony            0.010753   
32                  Pr. Accountant General, Teynampet            0.010309   
33   Raja Annamalaipuram, Ramakrishna Nagar (Chennai)            0.000000   
34                                Rajbhavan (Chennai)            0.010309   
35                                          Rayapuram            0.000000   
36                          Royapettah, Lloyds Estate            0.012821   
37                   Saidapet (Chennai), Guindy North            0.000000   
38                                         Saligramam            0.000000   
39  Shastri Nagar (Chennai), Kasturibai Nagar, Ady...            0.000000   
40                          Shenoy Nagar, Aminjikarai            0.000000   
41  Sowcarpet, Mannady (Chennai), Chennai G.P.O., ...            0.000000   
42   Teynampet West, DPI, Shastri Bhavan, Greams Road            0.000000   
43  Thygarayanagar North ND, Thygarayanagar South ...            0.000000   
44                          Tidel Park, TTTI Taramani            0.000000   
45     Tondiarpet, Tondiarpet West, Tondiarpet Bazaar            0.000000   
46                                         Vadapalani            0.000000   
47  Valmiki Nagar, Tiruvanmiyur, Palavakkam (Kanch...            0.000000   
48                                          Velacheri            0.000000   
49    Venkatesapuram, Puliyanthope, Perambur Barracks            0.000000   
50                                             Vepery            0.000000   
51                                      Vir

In [35]:
chennai_onehot_grouped.shape

(55, 149)

### Function is created to view the top 10 venues from each of the area

In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Places']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
chennai_venues_sorted = pd.DataFrame(columns=columns)
chennai_venues_sorted['Places'] = chennai_onehot_grouped['Places']

for ind in np.arange(chennai_onehot_grouped.shape[0]):
    chennai_venues_sorted.iloc[ind, 1:] = return_most_common_venues(chennai_onehot_grouped.iloc[ind, :], num_top_venues)

chennai_venues_sorted

Places 1st Most Common Venue  \
0                                Anna Nagar (Chennai)     Indian Restaurant   
1                                     Anna Nagar East     Indian Restaurant   
2                             Anna Nagar Western Extn     Indian Restaurant   
3                    Arumbakkam, D G Vaishnav College     Indian Restaurant   
4                                           Aynavaram     Indian Restaurant   
5                          Besantnagar, Rajaji Bhavan     Indian Restaurant   
6   Chepauk, Tiruvallikkeni, Parthasarathy Koil, M...     Indian Restaurant   
7                    Chetput, World University Centre     Indian Restaurant   
8   Chintadripet, Anna Road H.O, Madras Electricit...     Indian Restaurant   
9                                         Choolaimedu     Indian Restaurant   
10                   Egmore, Ethiraj Salai, Egmore ND     Indian Restaurant   
11            Ekkaduthangal, Guindy Industrial Estate     Indian Restaurant   
12                      Engineering College (Chennai)     Indian Restaurant   
13                Erukkancheri, Kodungaiyur, Rv Nagar     Indian Restaurant   
14                                       Flowers Road     Indian Restaurant   
15                                     Fort St George     Indian Restaurant   
16                              Gopalapuram (Chennai)     Indian Restaurant   
17                      High Court Building (Chennai)     Indian Restaurant   
18                                         Icf Colony     Indian Restaurant   
19                     Indian Institute Of Technology     Indian Restaurant   
20                Jafferkhanpet, Ashoknagar (Chennai)     Indian Restaurant   
21                        Kalaignar Karunanidhi Nagar     Indian Restaurant   
22                   Kilpauk, Kilpauk Medical College     Indian Restaurant   
23                                        Kodambakkam     Indian Restaurant   
24                                        Kotturpuram     Indian Restaurant   
25  Madras Medical College, Park Town H.O, Edapala...     Indian Restaurant   
26  Mylapore H.O, Mandaveli, Vivekananda College M...     Indian Restaurant   
27                                           Nandanam     Indian Restaurant   
28  Nerkundram, Koyambedu Wholesale Market Com, Ko...     Indian Restaurant   
29  Nungambakkam, Nungambakkam High Road, Loyola C...     Indian Restaurant   
30                  Perambur, Sembiam, Perambur North     Indian Restaurant   
31         Periyar Nagar, Kumaran Nagar, G K M Colony     Indian Restaurant   
32                  Pr. Accountant General, Teynampet                 Hotel   
33   Raja Annamalaipuram, Ramakrishna Nagar (Chennai)     Indian Restaurant   
34                                Rajbhavan (Chennai)     Indian Restaurant   
35                                          Rayapuram     Indian Restaurant   
36                          Royapettah, Lloyds Estate     Indian Restaurant   
37                   Saidapet (Chennai), Guindy North     Indian Restaurant   
38                                         Saligramam     Indian Restaurant   
39  Shastri Nagar (Chennai), Kasturibai Nagar, Ady...     Indian Restaurant   
40                          Shenoy Nagar, Aminjikarai     Indian Restaurant   
41  Sowcarpet, Mannady (Chennai), Chennai G.P.O., ...     Indian Restaurant   
42   Teynampet West, DPI, Shastri Bhavan, Greams Road     Indian Restaurant   
43  Thygarayanagar North ND, Thygarayanagar South ...     Indian Restaurant   
44                          Tidel Park, TTTI Taramani     Indian Restaurant   
45     Tondiarpet, Tondiarpet West, Tondiarpet Bazaar     Indian Restaurant   
46                                         Vadapalani     Indian Restaurant   
47  Valmiki Nagar, Tiruvanmiyur, Palavakkam (Kanch...     Indian Restaurant   
48                                          Velacheri     Indian Restaurant   
49    Venkatesapuram, Puliyanthope, Perambur Barracks     Indian Restaurant   
50                

### Clustering the above data into groups to know about the similarities in areas.
### By this the most suitable area cluster can be identified and the others clusters can be ignored instead of checking them one by one.

In [38]:
kclusters = 5

chennai_onehot_grouped_clustering = chennai_onehot_grouped.drop('Places', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(chennai_onehot_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:] 

array([1, 2, 2, 2, 0, 1, 3, 0, 3, 1, 3, 0, 0, 3, 0, 3, 3, 2, 0, 0, 0, 1,
       4, 3, 0, 3, 3, 0, 2, 0, 4, 3, 0, 0, 0, 4, 3, 2, 3, 0, 2, 4, 3, 2,
       1, 4, 3, 0, 3, 2, 4, 3, 4, 4, 3], dtype=int32)

In [39]:
chennai_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_) #merging the cluster labels with the data

chennai_merged = chennai_data

chennai_merged = chennai_merged.join(chennai_venues_sorted.set_index('Places'), on='Places')

chennai_merged.head()

Pincode                                             Places  Latitude  \
0  600018                  Pr. Accountant General, Teynampet   13.0433   
1  600020  Shastri Nagar (Chennai), Kasturibai Nagar, Ady...   12.9967   
2  600082         Periyar Nagar, Kumaran Nagar, G K M Colony   13.0572   
4  600040                               Anna Nagar (Chennai)   12.8819   
5  600102                                    Anna Nagar East   13.0974   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0    80.2528               0                 Hotel     Indian Restaurant   
1    80.2603               0     Indian Restaurant        Ice Cream Shop   
2    80.2554               3     Indian Restaurant                 Hotel   
4    80.0885               1     Indian Restaurant                Market   
5    80.1950               2     Indian Restaurant  Fast Food Restaurant   

           3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                           Café                Lounge            Restaurant   
1                           Café                 Beach    Chinese Restaurant   
2                           Café             Multiplex        Clothing Store   
4                    Coffee Shop           Pizza Place   Indie Movie Theater   
5  Vegetarian / Vegan Restaurant             Multiplex                  Café   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Italian Restaurant        Ice Cream Shop    Chinese Restaurant   
1                      Hotel  Fast Food Restaurant          Dessert Shop   
2  Middle Eastern Restaurant         Movie Theater        Ice Cream Shop   
4                Zoo Exhibit           Sports Club        Sandwich Place   
5             Clothing Store           Bus Station                Bakery   

  9th Most Common Venue          10th Most Common Venue  
0              Tea Room                            Park  
1           Snack Place                       Multiplex  
2    Italian Restaurant                      Donut Shop  
4         Movie Theater  Multicuisine Indian Restaurant  
5           Pizza Place              Chinese Restaurant

### Visualizing the Chennai map with the clusters to see how they are grouped

In [40]:
# create map
cluster_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(chennai_merged['Latitude'], chennai_merged['Longitude'], chennai_merged['Places'], chennai_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(cluster_map)
       
cluster_map

### Analyzing each cluster to know about the most common places

### As analyzing as a whole data, it is clear that most of the areas have their top most common places to be Indian Restaurant.
### So we can ignore it and start analyzing from the 2nd most common place in each area.

In [41]:
c1 = chennai_merged.loc[chennai_merged['Cluster Labels'] == 0, chennai_merged.columns[[1] + list(range(5, chennai_merged.shape[1]))]]

In [42]:
c2 = chennai_merged.loc[chennai_merged['Cluster Labels'] == 1, chennai_merged.columns[[1] + list(range(5, chennai_merged.shape[1]))]]

In [43]:
c3 = chennai_merged.loc[chennai_merged['Cluster Labels'] == 2, chennai_merged.columns[[1] + list(range(5, chennai_merged.shape[1]))]]

In [44]:
c4 = chennai_merged.loc[chennai_merged['Cluster Labels'] == 3, chennai_merged.columns[[1] + list(range(5, chennai_merged.shape[1]))]]

In [45]:
c5 = chennai_merged.loc[chennai_merged['Cluster Labels'] == 4, chennai_merged.columns[[1] + list(range(5, chennai_merged.shape[1]))]]

In [56]:
c1

Places 1st Most Common Venue  \
0                   Pr. Accountant General, Teynampet                 Hotel   
1   Shastri Nagar (Chennai), Kasturibai Nagar, Ady...     Indian Restaurant   
9                 Jafferkhanpet, Ashoknagar (Chennai)     Indian Restaurant   
10                                          Aynavaram     Indian Restaurant   
15            Ekkaduthangal, Guindy Industrial Estate     Indian Restaurant   
18                   Chetput, World University Centre     Indian Restaurant   
22                      Engineering College (Chennai)     Indian Restaurant   
26                                       Flowers Road     Indian Restaurant   
27   Raja Annamalaipuram, Ramakrishna Nagar (Chennai)     Indian Restaurant   
35                                         Icf Colony     Indian Restaurant   
36                     Indian Institute Of Technology     Indian Restaurant   
41                                        Kotturpuram     Indian Restaurant   
45  Nungambakkam, Nungambakkam High Road, Loyola C...     Indian Restaurant   
48                                           Nandanam     Indian Restaurant   
51                                Rajbhavan (Chennai)     Indian Restaurant   
53  Valmiki Nagar, Tiruvanmiyur, Palavakkam (Kanch...     Indian Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Indian Restaurant                  Café                Lounge   
1         Ice Cream Shop                  Café                 Beach   
9                   Café        Ice Cream Shop    Chinese Restaurant   
10                 Hotel                  Café    Chinese Restaurant   
15        Ice Cream Shop                  Café    Chinese Restaurant   
18        Ice Cream Shop    Chinese Restaurant                  Café   
22                 Hotel                  Café    Chinese Restaurant   
26                  Café        Ice Cream Shop    Chinese Restaurant   
27            Restaurant                 Hotel                Bakery   
35        Ice Cream Shop                  Café    Chinese Restaurant   
36        Ice Cream Shop                  Café    Chinese Restaurant   
41                  Café                 Hotel    Italian Restaurant   
45        Ice Cream Shop    Chinese Restaurant                  Café   
48                 Hotel    Italian Restaurant      Department Store   
51                 Hotel                  Café    Chinese Restaurant   
53                 Beach                  Café          Burger Joint   

   5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0             Restaurant    Italian Restaurant             Ice Cream Shop   
1     Chinese Restaurant                 Hotel       Fast Food Restaurant   
9            Coffee Shop                 Hotel         Italian Restaurant   
10    Italian Restaurant        Ice Cream Shop  Middle Eastern Restaurant   
15            Restaurant    Italian Restaurant                Pizza Place   
18           Coffee Shop    Italian Restaurant       Fast Food Restaurant   
22    Italian Restaurant        Ice Cream Shop  Middle Eastern Restaurant   
26           Coffee Shop                 Hotel         Italian Restaurant   
27                  Café    Italian Restaurant                     Lounge   
35            Restaurant    Italian Restaurant                Pizza Place   
36            Restaurant    Italian Restaurant                Pizza Place   
41                Lounge    Chinese Restaurant           Department Store   
45           Coffee Shop      Asian Restaurant                  BBQ Joint   
48            Restaurant        Clothing Store                     Lounge   
51    Italian Restaurant        Ice Cream Shop  Middle Eastern Restaurant   
53  Fast Food Restaurant           Coffee Shop             Ice Cream Shop   

      8th Most Common Venue    9th Most Common Venue   10th Most Common Venue  
0        Chinese Restaurant                 Tea Room                     Park  
1       

### As we can see in the first cluster, most of the top 5 common places are Indian Restaurants, Chinese Restaurants, Fast Food Restaurants and Café.
### As Fast Food Restaurants also have Chinese cuisines we can choose to ignore them.
### And also there are less number of Multiplexes or theatres, park, zoo, bus or train stations. 
### It would be a good decision to ignore this cluster as a whole as it is against our choosing criteria.

In [57]:
c2

Places 1st Most Common Venue 2nd Most Common Venue  \
4          Anna Nagar (Chennai)     Indian Restaurant                Market   
11   Besantnagar, Rajaji Bhavan     Indian Restaurant                Market   
14    Tidel Park, TTTI Taramani     Indian Restaurant   Indie Movie Theater   
20                  Choolaimedu     Indian Restaurant                Market   
37  Kalaignar Karunanidhi Nagar     Indian Restaurant           Pizza Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4            Coffee Shop           Pizza Place   Indie Movie Theater   
11   Indie Movie Theater         Train Station  Fast Food Restaurant   
14           Sports Club                   Zoo                  Food   
20   Indie Movie Theater         Train Station  Fast Food Restaurant   
37  Fast Food Restaurant        Ice Cream Shop         Movie Theater   

             6th Most Common Venue 7th Most Common Venue  \
4                      Zoo Exhibit           Sports Club   
11                     Coffee Shop                  Café   
14  Multicuisine Indian Restaurant      Department Store   
20                     Coffee Shop                  Café   
37                            Café             Bookstore   

   8th Most Common Venue 9th Most Common Venue          10th Most Common Venue  
4         Sandwich Place         Movie Theater  Multicuisine Indian Restaurant  
11                   Zoo             Juice Bar  Multicuisine Indian Restaurant  
14           Pizza Place             Racetrack                     Coffee Shop  
20                   Zoo             Juice Bar  Multicuisine Indian Restaurant  
37      Department Store                   Pub                            Road

### As for the 2nd cluster, it has more crowded places like theatres, zoo and markets. So we'll take this into consideration

In [58]:
c3

Places 1st Most Common Venue  \
5                                     Anna Nagar East     Indian Restaurant   
6                             Anna Nagar Western Extn     Indian Restaurant   
8                    Arumbakkam, D G Vaishnav College     Indian Restaurant   
21    Venkatesapuram, Puliyanthope, Perambur Barracks     Indian Restaurant   
32                   Saidapet (Chennai), Guindy North     Indian Restaurant   
33                      High Court Building (Chennai)     Indian Restaurant   
34  Thygarayanagar North ND, Thygarayanagar South ...     Indian Restaurant   
42  Nerkundram, Koyambedu Wholesale Market Com, Ko...     Indian Restaurant   
49                          Shenoy Nagar, Aminjikarai     Indian Restaurant   

            2nd Most Common Venue          3rd Most Common Venue  \
5            Fast Food Restaurant  Vegetarian / Vegan Restaurant   
6            Fast Food Restaurant  Vegetarian / Vegan Restaurant   
8            Fast Food Restaurant  Vegetarian / Vegan Restaurant   
21                    Bus Station                  Train Station   
32           Fast Food Restaurant                         Bakery   
33           Fast Food Restaurant  Vegetarian / Vegan Restaurant   
34           Fast Food Restaurant                         Bakery   
42  Vegetarian / Vegan Restaurant             Chinese Restaurant   
49  Vegetarian / Vegan Restaurant                    Coffee Shop   

            4th Most Common Venue    5th Most Common Venue  \
5                       Multiplex                     Café   
6                       Multiplex                     Café   
8              Chinese Restaurant           Clothing Store   
21                           Café  South Indian Restaurant   
32  Vegetarian / Vegan Restaurant                     Café   
33                      Multiplex                     Café   
34  Vegetarian / Vegan Restaurant                     Café   
42           Fast Food Restaurant              Coffee Shop   
49             Chinese Restaurant                     Café   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5         Clothing Store           Bus Station                Bakery   
6         Clothing Store           Bus Station                Bakery   
8              Multiplex     Electronics Store         Shopping Mall   
21  Fast Food Restaurant             Racetrack                Bakery   
32            Restaurant    Chinese Restaurant           Coffee Shop   
33        Clothing Store           Bus Station                Bakery   
34            Restaurant    Chinese Restaurant           Coffee Shop   
42        Clothing Store                  Café                 Hotel   
49  Fast Food Restaurant                Bakery            Restaurant   

   9th Most Common Venue     10th Most Common Venue  
5            Pizza Place         Chinese Restaurant  
6            Pizza Place         Chinese Restaurant  
8                  Hotel           Department Store  
21   Sporting Goods Shop                 Theme Park  
32                  Park  Middle Eastern Restaurant  
33           Pizza Place         Chinese Restaurant  
34                  Park  Middle Eastern Restaurant  
42          Burger Joint                 Restaurant  
49        Clothing Store                  Bookstore

### The third one has more Fast food restaurants and also less number of crowded places. So this one can be ignored.

In [59]:
c4

Places 1st Most Common Venue  \
2          Periyar Nagar, Kumaran Nagar, G K M Colony     Indian Restaurant   
7   Chintadripet, Anna Road H.O, Madras Electricit...     Indian Restaurant   
17  Chepauk, Tiruvallikkeni, Parthasarathy Koil, M...     Indian Restaurant   
23  Madras Medical College, Park Town H.O, Edapala...     Indian Restaurant   
24                   Egmore, Ethiraj Salai, Egmore ND     Indian Restaurant   
25                Erukkancheri, Kodungaiyur, Rv Nagar     Indian Restaurant   
28                                     Fort St George     Indian Restaurant   
29                          Royapettah, Lloyds Estate     Indian Restaurant   
30                              Gopalapuram (Chennai)     Indian Restaurant   
31   Teynampet West, DPI, Shastri Bhavan, Greams Road     Indian Restaurant   
39                                        Kodambakkam     Indian Restaurant   
43                                      Virugambakkam     Indian Restaurant   
44                       West Mambalam, Mambalam R.S.     Indian Restaurant   
46  Mylapore H.O, Mandaveli, Vivekananda College M...     Indian Restaurant   
52                                         Saligramam     Indian Restaurant   
55                                         Vadapalani     Indian Restaurant   
56                                          Velacheri     Indian Restaurant   

   2nd Most Common Venue    3rd Most Common Venue  \
2                  Hotel                     Café   
7                  Hotel           Clothing Store   
17                 Hotel                     Café   
23        Clothing Store                    Hotel   
24                 Hotel                     Café   
25                  Café                    Hotel   
28        Clothing Store                    Hotel   
29                 Hotel                     Café   
30                 Hotel                     Café   
31        Clothing Store                    Hotel   
39        Clothing Store            Jewelry Store   
43                  Café                    Hotel   
44        Clothing Store  South Indian Restaurant   
46        Clothing Store                    Hotel   
52                  Café                    Hotel   
55                 Hotel           Clothing Store   
56                  Café                    Hotel   

            4th Most Common Venue          5th Most Common Venue  \
2                       Multiplex                 Clothing Store   
7                     Pizza Place      Middle Eastern Restaurant   
17                      Multiplex           Fast Food Restaurant   
23                           Café      Middle Eastern Restaurant   
24             Italian Restaurant                 Ice Cream Shop   
25                      Multiplex      Middle Eastern Restaurant   
28                           Café      Middle Eastern Restaurant   
29                     Restaurant                            Bar   
30                      Multiplex                     Restaurant   
31                           Café      Middle Eastern Restaurant   
39  Vegetarian / Vegan Restaurant                 Ice Cream Shop   
43                      Multiplex      Middle Eastern Restaurant   
44                  Jewelry Store                      BBQ Joint   
46                           Café      Middle Eastern Restaurant   
52                      Multiplex      Middle Eastern Restaurant   
55                 Ice Cream Shop  Vegetarian / Vegan Restaurant   
56                      Multiplex      Middle Eastern Restaurant   

        6th Most Common Venue    7th Most Common Venue  \
2   Middle Eastern Restaurant            Movie Theater   
7              Sandwich Place                     Café   
17  Middle Eastern Restaurant           Clothing Store   
23                Pizza Place           Sandwich Place   
24         Chinese Restaurant              Pizza Place   
25              Movie Theater       Chinese Restaurant   
28                Pizza Place         

### This one doesn't have most of the people going place but has less number of Fast food or Chinese restaurants. We can add this.

In [60]:
c5

Places 1st Most Common Venue  \
13                    Washermanpet, Washermanpet East     Indian Restaurant   
16  Sowcarpet, Mannady (Chennai), Chennai G.P.O., ...     Indian Restaurant   
38                                          Rayapuram     Indian Restaurant   
40                   Kilpauk, Kilpauk Medical College     Indian Restaurant   
50                  Perambur, Sembiam, Perambur North     Indian Restaurant   
54     Tondiarpet, Tondiarpet West, Tondiarpet Bazaar     Indian Restaurant   
57                                             Vepery     Indian Restaurant   
58                    Vyasarpadi, Vyasar Nagar Colony     Indian Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13         Train Station                 Hotel             Multiplex   
16                 Hotel           Pizza Place    Italian Restaurant   
38         Train Station    Italian Restaurant                 Hotel   
40                Bakery            Restaurant           Coffee Shop   
50                Bakery           Coffee Shop                  Café   
54                 Hotel         Train Station             Multiplex   
57                 Hotel             Multiplex                  Café   
58           Coffee Shop    Italian Restaurant         Train Station   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13        Farmers Market           Coffee Shop        Sandwich Place   
16        Farmers Market            Donut Shop             Multiplex   
38        Farmers Market             Multiplex     Electronics Store   
40        Ice Cream Shop        Clothing Store      Department Store   
50         Train Station             Multiplex  Fast Food Restaurant   
54        Farmers Market             Nightclub           Pizza Place   
57           Snack Place        Sandwich Place  Fast Food Restaurant   
58                 Hotel             Multiplex        Farmers Market   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13                Bus Station             Nightclub             Restaurant  
16                     Museum              Platform            Coffee Shop  
38                     Market           Snack Place                  Motel  
40                       Café  Fast Food Restaurant     Chinese Restaurant  
50  Middle Eastern Restaurant      Department Store             Restaurant  
54                     Museum                Market            Coffee Shop  
57         Seafood Restaurant              Platform     Italian Restaurant  
58       Fast Food Restaurant                  Food                 Market

### The last cluster is also favourable to our criteria. So we can add this cluster.

In [62]:
final_df = pd.concat([c2, c4, c5], axis=0)


In [63]:
final_df.shape

(30, 11)

### Counting the most number of crowded places in the top 10 places in each area and also taking into account the places to be avoided.

In [64]:
final_df["Key counts"] = np.nan

In [65]:
key_words = ['Train', 'Multiplex', 'Theatre', 'Bus', 'Clothing Store', 'Park', 'Zoo', 'Market', 'Mall'] #for most crowded areas
not_key_words = ['Fast Food', 'Chinese'] #to avoid Chinese and fast food restaurants

In [66]:
final_df.head()

Places 1st Most Common Venue  \
4                                Anna Nagar (Chennai)     Indian Restaurant   
11                         Besantnagar, Rajaji Bhavan     Indian Restaurant   
14                          Tidel Park, TTTI Taramani     Indian Restaurant   
20                                        Choolaimedu     Indian Restaurant   
37                        Kalaignar Karunanidhi Nagar     Indian Restaurant   
2          Periyar Nagar, Kumaran Nagar, G K M Colony     Indian Restaurant   
7   Chintadripet, Anna Road H.O, Madras Electricit...     Indian Restaurant   
17  Chepauk, Tiruvallikkeni, Parthasarathy Koil, M...     Indian Restaurant   
23  Madras Medical College, Park Town H.O, Edapala...     Indian Restaurant   
24                   Egmore, Ethiraj Salai, Egmore ND     Indian Restaurant   
25                Erukkancheri, Kodungaiyur, Rv Nagar     Indian Restaurant   
28                                     Fort St George     Indian Restaurant   
29                          Royapettah, Lloyds Estate     Indian Restaurant   
30                              Gopalapuram (Chennai)     Indian Restaurant   
31   Teynampet West, DPI, Shastri Bhavan, Greams Road     Indian Restaurant   
39                                        Kodambakkam     Indian Restaurant   
43                                      Virugambakkam     Indian Restaurant   
44                       West Mambalam, Mambalam R.S.     Indian Restaurant   
46  Mylapore H.O, Mandaveli, Vivekananda College M...     Indian Restaurant   
52                                         Saligramam     Indian Restaurant   
55                                         Vadapalani     Indian Restaurant   
56                                          Velacheri     Indian Restaurant   
13                    Washermanpet, Washermanpet East     Indian Restaurant   
16  Sowcarpet, Mannady (Chennai), Chennai G.P.O., ...     Indian Restaurant   
38                                          Rayapuram     Indian Restaurant   
40                   Kilpauk, Kilpauk Medical College     Indian Restaurant   
50                  Perambur, Sembiam, Perambur North     Indian Restaurant   
54     Tondiarpet, Tondiarpet West, Tondiarpet Bazaar     Indian Restaurant   
57                                             Vepery     Indian Restaurant   
58                    Vyasarpadi, Vyasar Nagar Colony     Indian Restaurant   

   2nd Most Common Venue    3rd Most Common Venue  \
4                 Market              Coffee Shop   
11                Market      Indie Movie Theater   
14   Indie Movie Theater              Sports Club   
20                Market      Indie Movie Theater   
37           Pizza Place     Fast Food Restaurant   
2                  Hotel                     Café   
7                  Hotel           Clothing Store   
17                 Hotel                     Café   
23        Clothing Store                    Hotel   
24                 Hotel                     Café   
25                  Café                    Hotel   
28        Clothing Store                    Hotel   
29                 Hotel                     Café   
30                 Hotel                     Café   
31        Clothing Store                    Hotel   
39        Clothing Store            Jewelry Store   
43                  Café                    Hotel   
44        Clothing Store  South Indian Restaurant   
46        Clothing Store                    Hotel   
52                  Café                    Hotel   
55                 Hotel           Clothing Store   
56                  Café                    Hotel   
13         Train Station                    Hotel   
16                 Hotel              Pizza Place   
38         Train Station       Italian Restaurant   
40                Bakery               Restaurant   
50                Bakery              Coffee Shop   
54                 Hotel            Train Station   
57                 Hotel                Multiplex   
58     

In [67]:
# to fill the Key counts column which decides the area with most crowded and least Chinese and Fast Food restaurants.
sum = 0
for i in range(len(final_df)):
    for j in range(10):
        venue = final_df.iloc[i, j+1]
        if (any(ele in venue for ele in key_words) & any(ele not in venue for ele in not_key_words)):
            sum = sum+1
    final_df.iloc[i,11] = sum
    sum = 0
final_df.head()

Places 1st Most Common Venue  \
4                                Anna Nagar (Chennai)     Indian Restaurant   
11                         Besantnagar, Rajaji Bhavan     Indian Restaurant   
14                          Tidel Park, TTTI Taramani     Indian Restaurant   
20                                        Choolaimedu     Indian Restaurant   
37                        Kalaignar Karunanidhi Nagar     Indian Restaurant   
2          Periyar Nagar, Kumaran Nagar, G K M Colony     Indian Restaurant   
7   Chintadripet, Anna Road H.O, Madras Electricit...     Indian Restaurant   
17  Chepauk, Tiruvallikkeni, Parthasarathy Koil, M...     Indian Restaurant   
23  Madras Medical College, Park Town H.O, Edapala...     Indian Restaurant   
24                   Egmore, Ethiraj Salai, Egmore ND     Indian Restaurant   
25                Erukkancheri, Kodungaiyur, Rv Nagar     Indian Restaurant   
28                                     Fort St George     Indian Restaurant   
29                          Royapettah, Lloyds Estate     Indian Restaurant   
30                              Gopalapuram (Chennai)     Indian Restaurant   
31   Teynampet West, DPI, Shastri Bhavan, Greams Road     Indian Restaurant   
39                                        Kodambakkam     Indian Restaurant   
43                                      Virugambakkam     Indian Restaurant   
44                       West Mambalam, Mambalam R.S.     Indian Restaurant   
46  Mylapore H.O, Mandaveli, Vivekananda College M...     Indian Restaurant   
52                                         Saligramam     Indian Restaurant   
55                                         Vadapalani     Indian Restaurant   
56                                          Velacheri     Indian Restaurant   
13                    Washermanpet, Washermanpet East     Indian Restaurant   
16  Sowcarpet, Mannady (Chennai), Chennai G.P.O., ...     Indian Restaurant   
38                                          Rayapuram     Indian Restaurant   
40                   Kilpauk, Kilpauk Medical College     Indian Restaurant   
50                  Perambur, Sembiam, Perambur North     Indian Restaurant   
54     Tondiarpet, Tondiarpet West, Tondiarpet Bazaar     Indian Restaurant   
57                                             Vepery     Indian Restaurant   
58                    Vyasarpadi, Vyasar Nagar Colony     Indian Restaurant   

   2nd Most Common Venue    3rd Most Common Venue  \
4                 Market              Coffee Shop   
11                Market      Indie Movie Theater   
14   Indie Movie Theater              Sports Club   
20                Market      Indie Movie Theater   
37           Pizza Place     Fast Food Restaurant   
2                  Hotel                     Café   
7                  Hotel           Clothing Store   
17                 Hotel                     Café   
23        Clothing Store                    Hotel   
24                 Hotel                     Café   
25                  Café                    Hotel   
28        Clothing Store                    Hotel   
29                 Hotel                     Café   
30                 Hotel                     Café   
31        Clothing Store                    Hotel   
39        Clothing Store            Jewelry Store   
43                  Café                    Hotel   
44        Clothing Store  South Indian Restaurant   
46        Clothing Store                    Hotel   
52                  Café                    Hotel   
55                 Hotel           Clothing Store   
56                  Café                    Hotel   
13         Train Station                    Hotel   
16                 Hotel              Pizza Place   
38         Train Station       Italian Restaurant   
40                Bakery               Restaurant   
50                Bakery              Coffee Shop   
54                 Hotel            Train Station   
57                 Hotel                Multiplex   
58     

In [68]:
final_df[final_df['Key counts']>3]

Places 1st Most Common Venue  \
13                 Washermanpet, Washermanpet East     Indian Restaurant   
38                                       Rayapuram     Indian Restaurant   
54  Tondiarpet, Tondiarpet West, Tondiarpet Bazaar     Indian Restaurant   
58                 Vyasarpadi, Vyasar Nagar Colony     Indian Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13         Train Station                 Hotel             Multiplex   
38         Train Station    Italian Restaurant                 Hotel   
54                 Hotel         Train Station             Multiplex   
58           Coffee Shop    Italian Restaurant         Train Station   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13        Farmers Market           Coffee Shop        Sandwich Place   
38        Farmers Market             Multiplex     Electronics Store   
54        Farmers Market             Nightclub           Pizza Place   
58                 Hotel             Multiplex        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
13           Bus Station             Nightclub             Restaurant   
38                Market           Snack Place                  Motel   
54                Museum                Market            Coffee Shop   
58  Fast Food Restaurant                  Food                 Market   

    Key counts  
13         4.0  
38         4.0  
54         4.0  
58         4.0

### Finally the dataframe is narrowed down to four areas in Chennai which are the best suitable for the client to start a Chinese Restaurant.

### Washermanpet area would be the best according to the data as it has Train Station, Multiplex and Market in the top 5. And also the bus station when considering top 10.